### DISCUSSION TITLE
Inconsistent Grid Control Remote Option

### DISCUSSION THREAD

**Discussion by fxstein** [wills106/homeassistant-solax-modbus#618] [General]
Hi,

First of thanks for this great integration!

I have a problem with the Remote Control Option: Grid Control

The setup: 3x10kW Gen4 Hybrid Solax with 4 batteries each. Inverter 1 set to Master. Tested to charge and discharge at 30kW through the Inverter modes. When trying to do the same through remote control, the power intake for charging varies greatly and swings between ⅓ to ⅔ of the requested Power.

![Image 11-12-23 at 7 12 PM](https://github.com/wills106/homeassistant-solax-modbus/assets/773967/f03397ee-3731-4cda-98f0-9dfc09c3eb8d)
![Image 11-12-23 at 7 51 PM](https://github.com/wills106/homeassistant-solax-modbus/assets/773967/ec186a4e-8fcb-4ed9-8978-485df8202063)

Here are the settings:

![Image 11-12-23 at 7 48 PM](https://github.com/wills106/homeassistant-solax-modbus/assets/773967/f5a969ff-38b3-44bf-8d52-49e905d3bbbb)
![Image 11-12-23 at 7 48 PM](https://github.com/wills106/homeassistant-solax-modbus/assets/773967/90ca2e2c-4609-4b12-9867-f3182c84aac5)

Am I missing a setting somewhere?

**Comment by infradom** [comment]
As most of us do not hava a master-slave setup, it will be difficult to analyze this. Documentation of the remotecontrol/modbus power mechanism is almost unexisting afaik.
Do the other remotecontrol modes work beter in your master-slave setup? I guess not.
As the enbabled_grid_mode target is computed from the houseload and measured grid power, maybe that computation is not correct in a master-slave setup.

(value_function_remotecontrol_recompute in plugin_solax.py)

**Reply by fxstein** [reply]
Thank you! Battery hold with Battery Control at 0 target works well.

Looking at the plugin_solax.py I think I can see what is going wrong: in Parent/Child mode inverter 1 seems to report weird or incorrect house load. with the charging load at least partially being shown as house load (maybe inverters 2 and 3 being incorrectly counted as house load). As such the charging load is being reduced too much in the recompute function. Need to play a bit with that.

**Reply by fxstein** [reply]
Yes that is the issue:

When in Master/Slave mode the Master - Inverter 1 reports the charging loads of Inverters 2 & 3 as houseload. When you sum up the 3 individual houseloads you get the correct true houseload. Inverter reports positive high load, 2 & 3 report negative house load. Put together they even each other out to the correct overall houseload. However the recompute plugin only sees the housload of inverter 1. So it starts at the correct value of 30kW, but 20 seconds later it drops to around 10kW (because it assumes 20kW houseload), 20 sec later it goes back up to 27kW, ... and swings as long as it take for houseload and charge loads to equal out. In the end I get about 17-19kW charge load.

**Comment by gerd091986** [comment]
Sorry slightly off topic, but are you connecting to your inverter through the pocket wifi or the COM/LCD port to modbus adapter?

**Reply by fxstein** [reply]
Pocket WiFi - one per inverter. Inverters at address 1, 2, 3 respectively. Updating at 23, 27 and 31 second intervals. (to minimize modbus conflicts). Only inverter 1 (master) is used to remote control, 2 and 3 are only used to read out sensors.

**Reply by gerd091986** [reply]
Thank you so much, you have given me an idea for an issue I have.

**Reply by gerd091986** [reply]
This reply helped me fix the issue I had, thank you !

**Reply by fxstein** [reply]
Glad it helped!

**Comment by fxstein** [comment]
Still looking for a solution. Not making much progress.

**Comment by fxstein** [comment]
Ok found the issue and a temporary workaround. Was going in circles not realizing I had to restart home assistant after making code changes in `plugin_solax.py`

In `Enabled Power Control` mode (the one with the least amount of calculations happening)

```
def value_function_remotecontrol_recompute(initval, descr, datadict):
...

# This is the problem in master/slave as houseload_brut is incorrect by Solax - it counts all slave inverters as houseload
ap_target = min(ap_target,  import_limit - houseload_brut)
```

Quick and dirty fix:

```
def value_function_remotecontrol_recompute(initval, descr, datadict):
...

# Problem averted ;-)
# ap_target = min(ap_target,  import_limit - houseload_brut)
```

Even if the houseload_brut came back correctly from Solax, it raises another issue: import_limit seems to be limited to the sum of inverters at the moment. That is an artificial limit by the integration. In reality the import_limit is defined by the main breakers and is often or probably always higher than that. 

When charging the batteries from grid at times of very low electricity rates, the need to charge as quickly as possible within an hour is critical. Would be great if the integration could push up the limit to at least 1.5x that of the inverters to allow for a decent amount of houseload in addition to charging. Of course caution is required as the inverters will ignore the main breaker limit in its configuration when in remote control mode.

**Reply by fxstein** [reply]
For reference here the load stats during `Enable Power Control` mode with 3 x 10kW inverters charging at 30kW

![Image 11-21-23 at 10 07 PM](https://github.com/wills106/homeassistant-solax-modbus/assets/773967/b99deef8-18fe-47e5-89f5-80f742b585e1)

**Comment by fxstein** [comment]
And this is a more permanent solution:

```
def value_function_remotecontrol_recompute(initval, descr, datadict):
...

# ap_target = min(ap_target,  import_limit - houseload_brut)
active_power_real = datadict.get('active_power_real', 0)
houseload_real = (meas + active_power_real) * -1
ap_target = min(ap_target,  import_limit - houseload_real)

```

**Reply by cleveHEX** [reply]
If you think this will solve it permanently and won't regress for anyone, you can open a Pull Request to mainline the changes.

**Reply by fxstein** [reply]
Planning to do just that - but the big issue as you state is regression for other use cases. We will need individuals on single inverters to test that. About to get a single inverter installed at my sisters house and will test there as well but that is probably 6 weeks out.

**Reply by cleveHEX** [reply]
I could test it on a X3 Hybrid G4, but I do not have any experience with remotecontrol, so I don't know what to test.

**Reply by infradom** [reply]
I will test it also on my X3 Hybrid G4, I do use remotecontrol almost all the time.
I may notbe able to test it before the weekend however

**Reply by infradom** [reply]
Sorry for the delay, I did not forget, too busy with a renovation project. I will try to test it in the coming days ...

**Reply by fxstein** [reply]
Ok - My work-schedule has changed as such that I will have time to finally get this contributed.
